### ISSUE TITLE
Windows Node cannot pull images from private registry

### ISSUE THREAD

**Issue by luisxkimo** [rancher/rke2#7103]
**Environmental Info:**
RKE2 Version: 
rke2 version v1.30.5+rke2r1 (0c83bc82315cd61664880d0b52a7e070e9fbd623)
go version go1.22.6 X:boringcrypto

Node(s) CPU architecture, OS, and Version: 
Windows Server 2022 21H2 Build 20348.2700

Cluster Configuration:
2 Managers and 2 workers in RHEL all of them

**Describe the bug:**
Windows Node cannot pull images from private registry.

**Steps To Reproduce:**
- Create a pod using private registry in the name of the image

**Expected behavior:**
Image is downloaded and pod is running

**Actual behavior:**
Pod is not running with an error pulling the container image

**Additional context / logs:**
Here is an example of "describe" log of the pod:

```
Events:
  Type     Reason     Age   From               Message
  ----     ------     ----  ----               -------
  Normal   Scheduled  3s    default-scheduler  Successfully assigned kube-system/csi-proxy-mzlvg to win-node.internal.ad.com
  Normal   Pulling    2s    kubelet            Pulling image "ghcr.internal-cache.com/kubernetes-sigs/sig-windows/csi-proxy:v1.1.2"
  Warning  Failed     2s    kubelet            Failed to pull image "ghcr.internal-cache.com/kubernetes-sigs/sig-windows/csi-proxy:v1.1.2": failed to pull and unpack image "ghcr.internal-cache.com/kubernetes-sigs/sig-windows/csi-proxy:v1.1.2": failed to resolve reference "ghcr.internal-cache.com/kubernetes-sigs/sig-windows/csi-proxy:v1.1.2": unexpected status from HEAD request to https://ghcr.internal-cache.com/v2/kubernetes-sigs/sig-windows/csi-proxy/manifests/v1.1.2: 403
  Warning  Failed     2s    kubelet            Error: ErrImagePull
  Normal   BackOff    2s    kubelet            Back-off pulling image "ghcr.internal-cache.com/kubernetes-sigs/sig-windows/csi-proxy:v1.1.2"
```

Here is the configuration of the file `C:\etc\rancher\rke2\registries.yaml`:

```yaml
configs:
  docker.internal-cache.com:
    auth:
      password: thPassword
      username: username2
  ghcr.internal-cache.com:
    auth:
      password: ghcrpassWord
      username: usernameGHCR
```

**Comment by brandond** [comment]
Are you sure the credentials are correct? Are you sure that the image exists on that registry?

Can you pull that image that image successfully if you do `ctr -n k8s.io image pull --user USER:PASSWORD ghcr.internal-cache.com/kubernetes-sigs/sig-windows/csi-proxy:v1.1.2` ?

**Comment by luisxkimo** [comment]
Hi @brandond ,

Yes, the command to pull manually the image works fine using the same credentials on the `registries.yaml` file.

I can't find any kind of `containerd.log` inside `C:\var\log` and subfolders, but in any case, seeing the error with "403" in the events of the pod, I guess that the issue is related with the credentials that is trying to use.

Maybe the issue is that I haven't the right `registries.yaml` file or are in wrong path. Actually is `C:\etc\rancher\rke2\registries.yaml`

**Comment by brandond** [comment]
That should be the right path. Are there any errors in the system log regarding the contents of that file? Do you see the creds in `c:/var/lib/rancher/rke2/agent/etc/containerd/config.toml`?

**Comment by luisxkimo** [comment]
Yes, I can see a section on this config.toml like:

```
[plugins."io.containerd.grpc.v1.cri".registry]
      config_path = "C:\\var\\lib\\rancher\\rke2\\agent\\etc\\containerd\\certs.d"
      
        [plugins."io.containerd.grpc.v1.cri".registry.configs.auth."docker.internal-cache.com"]
          username = "username2"
          password = "thPassword"
     
        [plugins."io.containerd.grpc.v1.cri".registry.configs.auth."ghcr.internal-cache.com"]
          username = "usernameGHCR"
          password = "ghcrpassWord"
```

**Comment by brandond** [comment]
That's correct then, and all that K3s is responsible for managing. Take a look at the containerd.log (also under the rke2 agent dir) and see what that says.

**Comment by brandond** [comment]
You can confirm that you're pulling ages from docker.internal-cache.com? It doesn't have a port in the image name, or something else that would make a strong comparison fail?

**Comment by luisxkimo** [comment]
Hi,

Yes I confirmed that the image is downloaded from the cache. 

Anyway, we finally abandon the approach of use Windows node because these kind of issues and other related with csi plugins.

So I will close the thread without a real solution of the original issue but we cannot continue tracking this by now.

Thanks a lot @brandond for your support.

**Comment by HarrisonWAffel** [comment]
I think this ticket should be reopened, or I can create a new one.

It looks the issue is the [template used](https://github.com/k3s-io/k3s/blob/b55aaeb08d4d54efce39eca0d475e4080b1e177c/pkg/agent/templates/templates_windows.go#L118-L125) by k3s to render the `config.toml` file onto the node is using [invalid syntax](https://github.com/containerd/containerd/blob/main/docs/cri/registry.md#configure-registry-credentials) when setting up the credentials. I was able to reproduce this issue and resolved it by creating a `config.toml.tmpl` file that is nearly identical to the one shipped in k3s, with the only change being where `.auth` is placed 

`[plugins."io.containerd.grpc.v1.cri".registry.configs.auth."{{$k}}"]` 

To 

`[plugins."io.containerd.snapshotter.v1.stargz".registry.configs."{{$k}}".auth]`

This [matches linux](https://github.com/k3s-io/k3s/blob/b55aaeb08d4d54efce39eca0d475e4080b1e177c/pkg/agent/templates/templates_linux.go#L90-L94) and properly pulls images from an authenticated registry. Looking through git blame it seems that this section of the template hasn't been updated since it was first added. I don't see any mention of this use case in rancher/rancher or any of the windows repositories, so I'm not sure if authenticated registries were ever tested properly on Windows (though I could be wrong, not sure if rke2/k3s team tested this out at some point). 

SURE-9200

**Comment by brandond** [comment]
@HarrisonWAffel In the linux template the auths are in there twice, one for cri and one for stargz. Windows doesn't support stargz snapshotter so they're only in there once, for cri.

Windows:
* cri:  https://github.com/k3s-io/k3s/blob/b55aaeb08d4d54efce39eca0d475e4080b1e177c/pkg/agent/templates/templates_windows.go#L118-L128
Linux:
* stargz: https://github.com/k3s-io/k3s/blob/b55aaeb08d4d54efce39eca0d475e4080b1e177c/pkg/agent/templates/templates_linux.go#L52-L62
* cri: https://github.com/k3s-io/k3s/blob/b55aaeb08d4d54efce39eca0d475e4080b1e177c/pkg/agent/templates/templates_linux.go#L87-L97

**Comment by HarrisonWAffel** [comment]
Oh my bad that's a copy paste error on my part, I've updated my comment. In my test I did properly use `plugins."io.containerd.grpc.v1.cri".registry.configs`, with the only change being the placement of `.auth`

**Comment by brandond** [comment]
Tracking in https://github.com/k3s-io/k3s/issues/11693

**Comment by mdrahman-suse** [comment]
Validated on all the RC versions v1.29.14-rc2, v1.30.10-rc2, v1.31.6-rc3, v1.32.2-rc4
